compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -795,12 +804,48 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                 let violations = self.tcx.object_safety_violations(did);
                 report_object_safety_error(self.tcx, span, did, violations)
             }
-
             SelectionError::NotConstEvaluatable(NotConstEvaluatable::MentionsInfer) => {
                 bug!(
                     "MentionsInfer should have been handled in `traits/fulfill.rs` or `traits/select/mod.rs`"
                 )
             }
+            SelectionError::NotConstEvaluatable(NotConstEvaluatable::Silent) => {
+                let pred = obligation.predicate.clone();
+
+                // Try to prove the obligation again without Reveal::Selection for
+                // better diagnostics
+                match pred.kind().skip_binder() {
+                    ty::PredicateKind::ConstEvaluatable(uv) => {
+                        self.report_const_eval_failure(
+                            uv,
+                            obligation.clone(),
+                            root_obligation,
+                            fallback_has_occurred,
+                        );
+                    }
+                    ty::PredicateKind::ConstEquate(c1, c2) => {
+                        if let Some(uv1) = c1.try_get_unevaluated() {
+                            self.report_const_eval_failure(
+                                uv1.shrink(),
+                                obligation.clone(),
+                                root_obligation,
+                                fallback_has_occurred,
+                            );
+                        }
+
+                        if let Some(uv2) = c2.try_get_unevaluated() {
+                            self.report_const_eval_failure(
+                                uv2.shrink(),
+                                obligation.clone(),
+                                root_obligation,
+                                fallback_has_occurred,
+                            );
+                        }
+                    }
+                    _ => {}
+                }
+                tcx.sess.struct_span_err(span, "failed to evaluate the given constant")
```

Comment:
all cases of this in the test suite are redundant. maybe make it a delay_span_bug?